### PR TITLE
Add support for ignite configuration file

### PR DESCRIPTION
**Pull request by darkowlzz**
This adds a new API type `Configuration` which is used to define the
default configurations of ignite. A Configuration object can be defined
in a config file as:

```yaml
apiVersion: ignite.weave.works/v1alpha3
kind: Configuration
metadata:
  name: test-config
spec:
  runtime: containerd
  networkPlugin: cni
  vmDefaults:
    memory: "2GB"
    diskSize: "3GB"
    cpus: 2
    sandbox:
      oci: weaveworks/ignite:dev
    kernel:
      oci: weaveworks/ignite-kernel:4.19.47
    ssh: true
```

By default, ignite looks for the config file at /etc/ignite/config.yaml.
If this file is found, ignite uses it to configure the ignite defaults.
`--ignite-config` flag can be used to explicitly pass a configuration
file.

- Changes the flag override behavior when using a VM config file. The flags are no longer ignored and override the VM configuration.
- Debug log for the configuration used:
Default config:
```console
$ sudo bin/ignite run weaveworks/ignite-ubuntu --name my-vm --ssh --log-level=debug
DEBU[0000] Using ignite default configurations
```
With global config:
```console
$ sudo bin/ignite run weaveworks/ignite-ubuntu --name my-vm --ssh --log-level=debug
DEBU[0000] Found default ignite configuration file /etc/ignite/config.yaml
DEBU[0000] Using ignite configuration file /etc/ignite/config.yaml
```
- Adds e2e tests to verify the behavior.
- Adds new document `docs/ignite-configuration.md` with examples of the ignite configuration.

Related to #315

### CODE REVIEW

**Comment by luxas** [comment]
I'd add this to a new v1alpha3 API version, in order to not break v1alpha2 compatibility. ref: https://github.com/weaveworks/ignite/issues/355

**Comment by darkowlzz** [comment]
Rebased on v1alpha3 API.

File: cmd/ignite/cmd/root.go [hunk 1]
```diff
@@ -115,9 +153,27 @@ func addGlobalFlags(fs *pflag.FlagSet) {
...
 // AddQuietFlag adds the quiet flag to a flagset
 func AddQuietFlag(fs *pflag.FlagSet) {
 	fs.BoolVarP(&logs.Quiet, "quiet", "q", logs.Quiet, "The quiet mode allows for machine-parsable output by printing only IDs")
 }
+
+// getConfigFromFile reads a config file and returns ignite configuration.
+func getConfigFromFile(configPath string) (*api.Configuration, error) {
+	componentConfig := &api.Configuration{}
+
+	if err := scheme.Serializer.DecodeFileInto(configPath, componentConfig); err != nil {
```
Comment (twelho) [inline cmd/ignite/cmd/root.go:168]: TODO: Fix libgitops `DecodeFileInto` to not allow empty files

File: pkg/apis/ignite/v1alpha3/types.go [hunk 2]
```diff
@@ -237,3 +239,19 @@ type VMStatus struct {
...
+	runtime.ObjectMeta `json:"metadata"`
+
+	Spec ConfigurationSpec `json:"spec"`
+}
+
+// ConfigurationSpec defines the ignite configuration.
+type ConfigurationSpec struct {
+	Runtime       igniteRuntime.Name       `json:"runtime,omitempty"`
+	NetworkPlugin igniteNetwork.PluginName `json:"networkPlugin,omitempty"`
+	VM            VMSpec                   `json:"vm,omitempty"`
```
Comment (twelho) [inline pkg/apis/ignite/v1alpha3/types.go:256]: Rename to VMDefaults ... `vmDefaults`

File: cmd/ignite/cmd/root.go [hunk 3]
```diff
@@ -52,6 +58,38 @@ func NewIgniteCommand(in io.Reader, out, err io.Writer) *cobra.Command {
...
+			} else {
+				// Check the default config location.
+				if _, err := os.Stat(constants.IGNITE_CONFIG_FILE); !os.IsNotExist(err) {
+					log.Infof("Found default ignite configuration file %s", constants.IGNITE_CONFIG_FILE)
+					configFilePath = constants.IGNITE_CONFIG_FILE
+				}
+			}
+
+			if configFilePath != "" {
+				log.Infof("Using ignite configuration file %s", configFilePath)
```
Comment (stealthybox) [inline cmd/ignite/cmd/root.go:77]: We decided to keep these logs:
- increase the log level so it's not noisy in Info
- also log when there is no config found at all

File: cmd/ignite/cmd/root.go [hunk 4]
```diff
@@ -52,6 +58,38 @@ func NewIgniteCommand(in io.Reader, out, err io.Writer) *cobra.Command {
...
+				configFilePath = configPath
+			} else {
+				// Check the default config location.
+				if _, err := os.Stat(constants.IGNITE_CONFIG_FILE); !os.IsNotExist(err) {
+					log.Infof("Found default ignite configuration file %s", constants.IGNITE_CONFIG_FILE)
+					configFilePath = constants.IGNITE_CONFIG_FILE
+				}
+			}
+
+			if configFilePath != "" {
```
Comment (twelho) [inline cmd/ignite/cmd/root.go:76]: Also handle the config for `ignited`?

File: e2e/config_test.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+				runArgs = append(runArgs, "--config="+vmConfigFileName)
+			}
+
+			// Append the args to the run args for override flags.
+			runArgs = append(runArgs, rt.args...)
+			runCmd := exec.Command(
+				igniteBin,
+				runArgs...,
+			)
+			runOut, _ := runCmd.CombinedOutput()
```
Comment (twelho) [inline e2e/config_test.go:192]: Catch exit code using the returned error

File: cmd/ignite/cmd/root.go (hunk 4 above)
Comment (darkowlzz) [reply cmd/ignite/cmd/root.go:75]: Maybe in a follow-up PR?

File: cmd/ignite/cmd/root.go [hunk 6]
```diff
@@ -115,9 +153,27 @@ func addGlobalFlags(fs *pflag.FlagSet) {
 	logflag.LogLevelFlagVar(fs, &logLevel)
 	runtimeflag.RuntimeVar(fs, &providers.RuntimeName)
 	networkflag.NetworkPluginVar(fs, &providers.NetworkPluginName)
+	fs.StringVar(&configPath, "ignite-config", "", "Ignite configuration path")
```
[the remark this answers is not in the document]
Comment (darkowlzz) [reply cmd/ignite/cmd/root.go:157]: I've added a document about this in `docs/ignite-configuration.md`. I hope this will make it more clear.
Should this still be more descriptive?

File: cmd/ignite/run/create_test.go [hunk 7]
```diff
@@ -120,53 +251,28 @@ func TestConstructVMFromCLI(t *testing.T) {
...
-			name: "with no VM name and --require-name flag set",
-			createFlag: &CreateFlags{
-				VM:          &api.VM{},
-				RequireName: true,
-			},
-			args: []string{testImage},
-			err:  true,
-		},
-		{
-			name: "with sandbox image",
```
Comment (twelho) [inline cmd/ignite/run/create_test.go:139]: This is already tested in e2e so it's fine to exclude it here (tricky to implement).

File: cmd/ignite/run/create_test.go [hunk 8]
```diff
@@ -120,53 +251,28 @@ func TestConstructVMFromCLI(t *testing.T) {
...
 				},
 			},
-			args: []string{testImage},
 			wantSSH: api.SSH{
 				Generate:  true,
 				PublicKey: "some-pub-key",
 			},
 		},
-		{
-			name: "with no VM name and --require-name flag set",
```
Comment (twelho) [inline cmd/ignite/run/create_test.go:130]: This was tested twice, check https://github.com/weaveworks/ignite/pull/601/files#diff-d0bf10b7da8bd0e055593597d507f004R306

File: cmd/ignite/run/create_test.go [hunk 9]
```diff
@@ -120,53 +251,28 @@ func TestConstructVMFromCLI(t *testing.T) {
...
-			args:             []string{testImage},
-			wantSandboxImage: sandboxOCIRef,
-		},
 	}
 
 	for _, rt := range tests {
 		t.Run(rt.name, func(t *testing.T) {
-			err := rt.createFlag.constructVMFromCLI(rt.args)
+			vm := rt.createFlag.VM
+			fs := flag.NewFlagSet("test", flag.ExitOnError)
```
Comment (twelho) [inline cmd/ignite/run/create_test.go:264]: This does not test flag validity since that would need the flag set to be populated (needs additional logic), but the flag handling will need to be refactored at some point anyway so this is fine.

File: cmd/ignite/run/create_test.go [hunk 10]
```diff
@@ -221,7 +322,23 @@ func TestNewCreateOptions(t *testing.T) {
 
 	for _, rt := range tests {
 		t.Run(rt.name, func(t *testing.T) {
-			_, err := rt.createFlag.NewCreateOptions([]string{})
+			// Setup storage backend.
+			dir, err := ioutil.TempDir("", "ignite")
+			if err != nil {
+				t.Fatalf("failed to create storage for ignite: %v", err)
+			}
+			defer os.RemoveAll(dir)
```
Comment (twelho) [inline cmd/ignite/run/create_test.go:330]: Catch this error with `util.DeferErr` (will be done in a separate PR)

File: cmd/ignite/run/create_test.go [hunk 11]
```diff
@@ -221,7 +322,23 @@ func TestNewCreateOptions(t *testing.T) {
...
+
+			storage := cache.NewCache(
```
Comment (twelho) [inline cmd/ignite/run/create_test.go:332]: This should use Ignite's configured providers to have consistency between the tests and built binary, can be done in a separate PR.
